x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java

```diff
@@ -297,9 +297,16 @@ private void executeCompletionListeners() {
             if (hasCompleted) {
                 return;
             }
-            hasCompleted = true;
-            completionsListenersCopy = new HashMap<>(this.completionListeners);
-            this.completionListeners.clear();
+            try {
+                hasCompleted = true;
+                completionsListenersCopy = new HashMap<>(this.completionListeners);
+                this.completionListeners.clear();
+            } finally {
+                if (searchResponse != null && searchResponse.get() != null) {
+                    MutableSearchResponse mutableSearchResponse = searchResponse.get();
+                    mutableSearchResponse.notifySearchTaskCompleted();
```

Comment:
Hmm, OK. I see what you mean.

I'm not sure I can see how to solve this. Don't we have a circular dependency here?

1) We don't want to mark AsyncSearchResponse with isRunning=false until the AsyncSearchTask has finished

2) But in order for the AsyncSearchTask to finish, its completion listeners have to get a "final" snapshot of the AsyncSearchResponse to store it the `.async-search` index. 

This [completion listener code](https://github.com/elastic/elasticsearch/blob/3b1c9236583fa0d96bb4f1b8b70d15f56d82b29d/x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java#L315) calls [TransportSubmitAsyncSearchAction.onFinalResponse](https://github.com/elastic/elasticsearch/blob/a6ca1e749224ee5117787e6ad29d9847574b1444/x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java#L208)  to store it.

But to accomplish goal 1, this "final" snapshot of AsyncSearchResponse written to the index will have `isRunning=true` if we wait to signal MutableSearchResponse that the task is finished only after the last completion listener exits.

What are the ways to solve this?

I've pushed another commit that basically has the completion listener get a copy of AsyncSearchResponse where isRunning has been overridden to false so it can store that one to disk, but any other caller asking for AsyncSearchResponse will still see isRunning=true until the final completion listener has finished.

If we don't like that - what other ideas are there?

Finally, while we do have a race condition here (in the code in main), does it really affect production code in any serious way? The issue that raised this is that IT tests hit this race condition and fail. So can the IT test wait until on something else besides isRunning=false flag in the response? Perhaps it could query the Task API and ensure the task is unregistered?

Reply:
With that commit, the tests revealed a new race condition between isPartial and isRunning. It was possible to get in the search response:
```
isRunning=true
isPartial=false
```
and the test was enforcing an invariant that if the search is still running then isPartial should be true.

My latest commit also dynamically adjusts isPartial based on whether the search is still running and if not, then it uses the final isPartial setting in `MutableSearchResponse`.

Reply:
This kind of lingering-activity-after-completion-notification thing is largely a problem for tests indeed, but that in itself makes it worth fixing. It does also highlight some design questions. For instance, does it make sense to record the `isRunning` flag in the index at all? What if the (async) action that writes the final snapshot fails? Or the cluster completely restarts mid-search?